content/api/servo-assembly/arduino-commands.md

```diff
@@ -0,0 +1,209 @@
+functionality you're adding to the Arduino, you'll need to define a new command.
+
+Adding a command requires two changes. First you must define the function which
+will implement the operations of that command and then you register that command
+under the name you'll use to call it from the Python code.
+
+### Implementing the command
+
+You'll need to define a function to implement your command. The function should
+be placed in the `firmware.cpp` file just above the definition of the `commands`
+array (likely about line 205).
+
+The function definition needs to look like this:
+
+```cpp
+static CommandError my_function(int commandId, String argument) {
+  // implementation here
+  return OK;
+}
+```
+
+You should choose a suitable name for your function (above it's called
+`my_function`), though you'll need to stick to letters, numbers and underscore
+(`_`). The rest of the function signature (the line starting `static
+CommandError` above) needs to be exactly as above.
+
+You command will need to adhere to some conventions to be able to communicate
+with the Pi, though you can otherwise do pretty much anything you normally would
+on an Arduino.
+
+The following example command shows how to handle arguments from the Pi as well
+as how to return errors and other values:
+
+```cpp
+static CommandError echo(int commandId, String argument) {
+
+  if (argument.length() == 0) {
+    // If something went wrong, you should return a `COMMAND_ERROR` with a
+    // message about the failure. That message will be fed back to Python where
+    // an exception will be raised.
+    return COMMAND_ERROR("nothing to echo");
+  }
+
+  while (argument.length() > 0) {
+    // You can use `pop_option` to extract individual parameters from the
+    // `argument` string. They're sent space-separated, so be sure not to
+    // include any spaces within your argument values!
+    //
+    // You don't need to check the length of `argument` before calling
+    // `pop_option` -- if there isn't an option to be had you'll get back an
+    // empty string ("").
+    String next_arg = pop_option(argument);
+
+    // You can write values to be read by the code on the Pi using the
+    // `serialWrite` function. The first two arguments need to be exactly as
+    // they are here; the third argument needs to be a `String` containing the
+    // value you want to send back to the Pi.
+    serialWrite(commandId, '>', next_arg);
+  }
+
+  // If everything went fine, you should return `OK`.
+  return OK;
+}
+```
+<!-- TODO: mention being able to send back comments? -->
+
+For other examples of how to define commands (and the helper functions you can
+use), you can have a look at the other commands the Arduino already supports.
+In particular the `get_version`, `led` and `servo` functions are a good starting
+point.
+
+<!-- TODO should we error in robot-api/robotd if the incoming arguments include spaces? -->
```

Comment:
This isn't really related to the docs - could we break this out into an issue on `robot-api` or a Trello card instead?

Reply:
2974f73